salt-mode.el

```diff
@@ -194,6 +213,202 @@ module to use.
+
+(defun salt-mode--query-minion (program)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+Error handling is the responsibility of the caller; failures
+may occur starting the Python process or parsing its output."
+  (with-temp-buffer
+    (process-file salt-mode-python-program nil t nil "-c"
+                  (format salt-mode--query-template program))
+    (goto-char (point-min))
+    (let ((json-array-type 'list)
+          (json-object-type 'hash-table))
+      (json-read))))
+
+(defun salt-mode--async-minion (program callback)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+JSON response data is passed to CALLBACK when it is ready. Errors
+result in a message but no signal."
+  (with-demoted-errors "Unable to query Salt minion: %s"
+    (message "starting in %s" default-directory)
+    (set-process-sentinel
+     (start-file-process "salt-async"
+                         ;; Work around TRAMP bug in TRAMP 2.2 / Emacs 25.1;
+                         ;; remote buffers cannot be passed by name.
+                         (get-buffer-create
+                          (generate-new-buffer-name " *salt-async*"))
+                         salt-mode-python-program "-c"
+                         (format salt-mode--query-template program))
+     (lambda (process _event)
+       (when (memq (process-status process) '(exit signal))
+         (with-current-buffer (process-buffer process)
+           (with-demoted-errors "Error querying Salt minion: %s"
+             (goto-char (point-min))
+             (condition-case nil
+                 (let ((json-array-type 'list)
+                       (json-object-type 'hash-table))
+                   (funcall callback (json-read)))
+               ((json-readtable-error)
+                (message "Error querying Salt minion: %s"
+                         (string-trim-right (buffer-string))))))
+           (kill-buffer)))))))
+
+(defmacro salt-mode--with-async-minion (program &rest body)
+  "Run Python code PROGRAM on a virtual Salt minion.
+
+BODY will run once the result is available, in the variable `result'."
+  (declare (indent 1))
+  `(salt-mode--async-minion ,program (lambda (result) ,@body)))
+
+
+(defvar salt-mode--state-argspecs nil
+  "Information about Salt states.")
+
+(defun salt-mode-refresh-data (&optional if-missing)
+  "Refresh the information about available Salt states.
+
+When IF-MISSING is set, only refresh data that is empty."
+  (interactive)
+  (unless (and salt-mode--state-argspecs if-missing)
+    (salt-mode--with-async-minion "minion.functions.sys.state_argspec('*')"
+      (when (and result (hash-table-p result))
+        (setq salt-mode--state-argspecs result)
+        (message "Loaded %d Salt state function argument specifications."
+                 (hash-table-count result))))))
+
+(defun salt-mode--state-doc (module-or-function)
+  "Return documentation for the given state MODULE-OR-FUNCTION."
+  (let ((doc (gethash module-or-function
```

Comment:
Two reasons:

- Concern about memory usage, maybe unnecessary concern. The argspecs only have the arguments, not the full documentation, which is much larger. (And even larger still if we go on to handle commands.)

- Virtual modules implementing common interfaces (e.g. `pkg`) differ less in arguments than detailed documentation. Sharing the argspecs between all buffers (which might be on different systems) is less fraught than sharing full documentation. I would rather share nothing, but I also don't like the overhead of separate cached argspecs for each buffer, or the overhead of uncached argspecs and going to the "minion" for each eldoc call. (And you're probably editing on a master with minimal relation to the minion configuration anyway, but, small victories.)